File: src/main/java/sx/jms/SpringJmsListener.java
package sx.jms;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.springframework.jms.connection.JmsTransactionManager;
import org.springframework.jms.listener.DefaultMessageListenerContainer;
import org.springframework.jms.listener.SessionAwareMessageListener;
import org.springframework.util.ErrorHandler;

import javax.jms.ConnectionFactory;
import javax.jms.Destination;
import javax.jms.JMSException;

/**
 * Spring listener implementation
 * Created by masc on 07.06.15.
 */
public abstract class SpringJmsListener extends Listener implements SessionAwareMessageListener, ErrorHandler {
    private Log mLog = LogFactory.getLog(this.getClass());

    /** Destination this listener is attached to */
    private Destination mDestination;
    /** Spring message listener container */
    private DefaultMessageListenerContainer mListenerContainer;
    /** Transaction manager */
    private JmsTransactionManager mTransactionManager;

    /**
     * c'tor
     * @param connectionFactory
     */
    public SpringJmsListener(ConnectionFactory connectionFactory) {
        super(connectionFactory);
    }

    /**
     * Creates listener destination
     * @return
     */
    protected abstract Destination createDestination();

    /**
     * Can be overridden to customize listener container configuration
     * @param listenerContainer
     */
    protected void configure(DefaultMessageListenerContainer listenerContainer) { }

    @Override
    public void start() throws JMSException {
        if (mListenerContainer == null) {
            mDestination = this.createDestination();

            mListenerContainer = new DefaultMessageListenerContainer();
            mListenerContainer.setConnectionFactory(this.getConnectionFactory());
            mListenerContainer.setMessageListener(this);
            mListenerContainer.setSessionTransacted(true);
            mListenerContainer.setErrorHandler(this);
            mListenerContainer.setDestination(mDestination);
            this.configure(mListenerContainer);
            mListenerContainer.afterPropertiesSet();
        }
        mListenerContainer.start();
    }

    @Override
    public void stop() throws JMSException {
        mListenerContainer.stop();
    }

    @Override
    public void handleError(Throwable t) {
        mLog.error(t.getMessage(), t);
    }

    @Override
    public void dispose() {
        try {
            this.stop();
        } catch (JMSException e) {
        }
    }
}

File: src/main/java/sx/mq/jms/listeners/SpringJmsListener.kt
package sx.mq.jms.listeners

import org.springframework.jms.connection.JmsTransactionManager
import org.springframework.jms.listener.DefaultMessageListenerContainer
import org.springframework.jms.listener.SessionAwareMessageListener
import org.springframework.util.ErrorHandler
import sx.mq.jms.JmsEndpoint
import sx.mq.jms.JmsListener
import java.util.concurrent.Executor
import javax.jms.JMSException
import javax.jms.Message
import javax.jms.Session

/**
 * Spring listener implementation
 * Created by masc on 07.06.15.
 * @param endpoint The channel to listen on
 * @param durationClientId The client id to use for durable topic subscriptions. Setting this identifier enables durable subscription.
 * @param executor The executor to spawn threads from
 */
open class SpringJmsListener(
        endpoint: JmsEndpoint,
        private val executor: Executor,
        private val durableClientId: String? = null)
    :
        JmsListener(endpoint),
        ErrorHandler {

    /** Spring message listener container  */
    private var listenerContainer: DefaultMessageListenerContainer? = null
    /** Transaction manager  */
    private val transactionManager: JmsTransactionManager? = null

    /**
     * Optionally overridden to customize listener container configuration
     * @param listenerContainer
     */
    open protected fun configure(listenerContainer: DefaultMessageListenerContainer) {
    }

    private val description by lazy({
        "[${this.javaClass.simpleName}] for [${this.endpoint}]"
    })

    /**
     * Start listener
     */
    @Synchronized final override fun start() {
        log.info("Starting ${this.description}")
        var lc = listenerContainer
        if (lc == null) {
            lc = DefaultMessageListenerContainer()

            if (this.durableClientId != null) {
                lc.clientId = this.durableClientId
                lc.isSubscriptionDurable = true
            }
            lc.connectionFactory = this.context.connectionFactory
            lc.messageListener = object : SessionAwareMessageListener<Message> {
                @Throws(JMSException::class)
                override fun onMessage(message: Message, session: Session) {
                    this@SpringJmsListener.onMessage(message, session)
                }
            }
            lc.isSessionTransacted = true
            lc.errorHandler = this
            lc.destination = this.endpoint.destination
            lc.setTaskExecutor(this.executor)
            this.configure(lc)
            lc.afterPropertiesSet()
            this.listenerContainer = lc
        }
        lc.start()
    }

    /**
     * Spring error handler
     * @param t Error
     */
    final override fun handleError(t: Throwable) {
        this.onError(t)
    }

    /**
     * Stop listener
     */
    @Synchronized override fun stop() {
        val lc = listenerContainer
        if (lc != null) {
            log.info("Stopping ${this.description}")
            lc.shutdown()
            listenerContainer = null
        }
    }
}
